src/Stratis.Bitcoin.Tests/Utilities/AsyncManualResetEventTest.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Stratis.Bitcoin.Utilities;
+using Xunit;
+
+namespace Stratis.Bitcoin.Tests.Utilities
+{
+    public class AsyncManualResetEventTest
+    {
+        /// <summary>
+        /// Tests that:
+        /// <list type="bullet">
+        /// <item><see cref="AsyncManualResetEvent"/> does not crush.</item>
+        /// <item>WaitAsync eventually completes.</item>
+        /// <item>WaitAsync completes when the event is set.</item>
+        /// </list>
+        /// </summary>
+        [Fact]
+        public async void AsyncManualResetEvent_WaitAsync()
+        {
+            AsyncManualResetEvent manualResetEvent = new AsyncManualResetEvent(false);
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
+
+            Task task = Task.Run(async () =>
+            {
+                await Task.Delay(500);
+                manualResetEvent.Set();
+            });
+
+            Task mreAwaitingTask = manualResetEvent.WaitAsync();
+            await mreAwaitingTask;
+
+            stopwatch.Stop();
+
+            Assert.True(stopwatch.ElapsedMilliseconds >= 500 && stopwatch.ElapsedMilliseconds < 600);
```

Comment:
this could blow as we've seen in my stopwatch tests
although you did allow 100 ms epsilon, so most of the time it should be ok, but better to avoid this
also use () or make two separate Asserts

Reply:
now i remembered that the stopwatch test is actually disabled on mac because it did not work even with higher epsilon

in general i'd prefer a test which instead of measuring time proves correctness by correct ordering, that would avoid problems with time measurement in test environment

so you can definitely replace this test with such a test of good ordering
so imagine two or more tasks (let's do two for simplicity, i'd do three for fun, but simplicity is better than fun) - A, B
and these guys want to cooperate to create a list of 100 integers from 1 to 100.

they will sync with this async event

so you create two events EA and EB, EA is set and EB is reset
you create two tasks over the same body and pass both events as the arguments, but for A, you pass events in order "EA, EB" and for B you pass events in order "EB, EA"

and you create shutdownEvent that is set to false

```
body(myEvent, theirEvent, shutdownEvent)
{
   loop until we are done
   {
       wait for myEvent + shutdownEvent
       reset myEvent
       get last number from the list
       add one to that number
       add the new number to the list
       if the newly added number is the limit to reach signal shutdownEvent
       set theirEvent
   }
}
```

the main thread should wait only for shutdownEvent and then asserts that the list is formed well

Reply:
and we could also assert that when A wakes up then EB is reset and vice versa

Reply:
I'm really not sure if we should overcomplicate this. 
Ok, I agree that we should the <600 condition just in case. But the ordering test is overkill. 

>proves correctness by correct ordering

This wasn't the intention of this test. This test tests if the event was triggered after the 500ms delay and not before, that's all.

Reply:
I know, but a little more complicated test would show that the async event is reliable. Maybe it is overkill and we can think of a much simpler test (although it is really 15 minutes of coding). 

The problem with this test is that it will possibly fail in some environments. The stopwatch tests that failed on Mac was more complex, so maybe this one would work, but conceptually, such measurement is not good and can fail (imagine your machine is under heavy stress, executing 200 threads and the thread that should jump of your task.delay won't get scheduled on time, the test will fail and it should not ...). So if you feel you can think out a simpler test, let's go for a simpler, but I'd like to avoid using stopwatch.

Reply:
Fixed. 

>so imagine two or more tasks (let's do two for simplicity, i'd do three for fun, but simplicity is better than fun)
Kept it simple. Had some fun. 
Now we have 10 tasks.